File: Web/src/utils/badgeDefinitions.ts
import { 
  Flame, 
  Zap, 
  Award, 
  Target, 
  Star, 
  Crown, 
  Sparkles, 
  Trophy,
  Rocket,
  Brain,
  BookOpen,
  TrendingUp,
  Shield,
  Gem,
  Key,
  Settings,
  Heart,
  Database,
  Sun,
  Moon,
  Sunrise,
  Layers,
  type LucideIcon
} from 'lucide-react';

export type BadgeCategory = 'streak' | 'mastery' | 'milestone';
export type BadgeRarity = 'common' | 'rare' | 'epic' | 'legendary';

export interface Badge {
  id: string;
  name: string;
  description: string;
  visual: string; // Visual description for designers
  category: BadgeCategory;
  rarity: BadgeRarity;
  icon: LucideIcon;
  color: string; // Gradient start color
  colorEnd: string; // Gradient end color
  requirement: string; // User-facing requirement text
  // Used for automatic badge awarding
  criteria?: {
    type: 'streak' | 'total_unlocks' | 'effort_score_avg' | 'mastery_unlocks' | 'mastery_mode' | 'perfect_score' | 'late_night' | 'early_morning' | 'saved_count' | 'streak_saved' | 'manual';
    value?: number;
  };
}

export const BADGE_DEFINITIONS: Badge[] = [
  // ===== STREAKS (6 Badges) =====
  // Theme: Maintaining the fire (Magma Orange → Blue Heat)
  {
    id: 'ignition',
    name: 'Ignition',
    description: 'The spark has been struck',
    visual: 'A single matchstick striking a spark against a dark surface',
    category: 'streak',
    rarity: 'common',
    icon: Flame,
    color: '#FF6B35',
    colorEnd: '#FF4500',
    requirement: '3-day streak',
    criteria: { type: 'streak', value: 3 },
  },
  {
    id: 'the_furnace',
    name: 'The Furnace',
    description: 'Heat is building, momentum is real',
    visual: 'An industrial furnace door glowing orange with heat leaking out',
    category: 'streak',
    rarity: 'rare',
    icon: Zap,
    color: '#FF8C42',
    colorEnd: '#FF6B35',
    requirement: '7-day streak',
    criteria: { type: 'streak', value: 7 },
  },
  {
    id: 'momentum',
    name: 'Momentum',
    description: 'The flywheel is spinning',
    visual: 'A heavy flywheel or turbine spinning with green motion blur',
    category: 'streak',
    rarity: 'rare',
    icon: TrendingUp,
    color: '#00FF94',
    colorEnd: '#00CC75',
    requirement: '14-day streak',
    criteria: { type: 'streak', value: 14 },
  },
  {
    id: 'blue_giant',
    name: 'Blue Giant',
    description: 'Burning with the hottest intensity',
    visual: 'A massive star burning with intense blue/white light (hottest flame)',
    category: 'streak',
    rarity: 'epic',
    icon: Star,
    color: '#00D9FF',
    colorEnd: '#0099CC',
    requirement: '30-day streak',
    criteria: { type: 'streak', value: 30 },
  },
  {
    id: 'the_century',
    name: 'The Century',
    description: 'Discipline forged in platinum',
    visual: 'The number "100" chiseled out of solid Platinum, glowing from behind',
    category: 'streak',
    rarity: 'legendary',
    icon: Trophy,
    color: '#E5E4E2',
    colorEnd: '#C0C0C0',
    requirement: '100-day streak',
    criteria: { type: 'streak', value: 100 },
  },
  {
    id: 'the_reboot',
    name: 'The Reboot',
    description: 'System restored, continuity maintained',
    visual: 'A "System Restore" icon or a Phoenix Feather digitized into pixels',
    category: 'streak',
    rarity: 'epic',
    icon: Shield,
    color: '#8A2BE2',
    colorEnd: '#6A1BB2',
    requirement: 'Used Streak Freeze',
    criteria: { type: 'streak_saved' },
  },

  // ===== MASTERY (7 Badges) =====
  // Theme: Depth of thought and quality (Electric Violet → Neon Pink)
  {
    id: 'synapse',
    name: 'Synapse',
    description: 'The connection has been made',
    visual: 'Two neural nodes connecting with a bright electrical zap',
    category: 'mastery',
    rarity: 'rare',
    icon: Zap,
    color: '#8A2BE2',
    colorEnd: '#A855F7',
    requirement: 'First high effort score',
    criteria: { type: 'effort_score_avg', value: 2.5 },
  },
  {
    id: 'deep_dive',
    name: 'Deep Dive',
    description: 'Explored the depths of understanding',
    visual: 'A futuristic diving helmet or a submarine window looking into a dark void',
    category: 'mastery',
    rarity: 'epic',
    icon: Target,
    color: '#3B82F6',
    colorEnd: '#2563EB',
    requirement: '100% on effort & understanding',
    criteria: { type: 'perfect_score' },
  },
  {
    id: 'vanguard',
    name: 'Vanguard',
    description: 'Elite difficulty conquered',
    visual: 'A tactical "Spec-Ops" helmet with a glowing red visor',
    category: 'mastery',
    rarity: 'epic',
    icon: Shield,
    color: '#DC2626',
    colorEnd: '#991B1B',
    requirement: 'Unlock in Mastery Mode',
    criteria: { type: 'mastery_mode' },
  },
  {
    id: 'the_architect',
    name: 'The Architect',
    description: 'Perfect logical structure detected',
    visual: 'A glowing wireframe cube (tesseract) floating in space',
    category: 'mastery',
    rarity: 'legendary',
    icon: Layers,
    color: '#00D9FF',
    colorEnd: '#8A2BE2',
    requirement: 'Perfect structure in explanation',
    criteria: { type: 'manual' }, // AI detects perfect structure
  },
  {
    id: 'the_polymath',
    name: 'The Polymath',
    description: 'Knowledge spans multiple domains',
    visual: 'A glass prism refracting a beam of light into a rainbow',
    category: 'mastery',
    rarity: 'legendary',
    icon: Sparkles,
    color: '#FF6B35',
    colorEnd: '#8A2BE2',
    requirement: '5 different subject categories',
    criteria: { type: 'manual' }, // Requires subject tracking
  },
  {
    id: 'night_shift',
    name: 'Night Shift',
    description: 'Late-night focus session completed',
    visual: 'A neon crescent moon wearing cyber-sunglasses',
    category: 'mastery',
    rarity: 'rare',
    icon: Moon,
    color: '#A855F7',
    colorEnd: '#EC4899',
    requirement: 'High effort after 11:00 PM',
    criteria: { type: 'late_night' },
  },
  {
    id: 'the_refiner',
    name: 'The Refiner',
    description: 'Crafted through iteration',
    visual: 'A scalpel or laser cutter working on a rough rock',
    category: 'mastery',
    rarity: 'epic',
    icon: Settings,
    color: '#10B981',
    colorEnd: '#059669',
    requirement: 'Revised explanation 3+ times',
    criteria: { type: 'manual' }, // Requires edit tracking
  },

  // ===== MILESTONES (7 Badges) =====
  // Theme: Experience levels and volume (Raw Materials → Precious Metals)
  {
    id: 'the_initiate',
    name: 'The Initiate',
    description: 'The journey begins',
    visual: 'A rough, unpolished stone key',
    category: 'milestone',
    rarity: 'common',
    icon: Key,
    color: '#78716C',
    colorEnd: '#57534E',
    requirement: '1 unlock',
    criteria: { type: 'total_unlocks', value: 1 },
  },
  {
    id: 'the_apprentice',
    name: 'The Apprentice',
    description: 'Learning the fundamentals',
    visual: 'A bronze gear, slightly rusted but functional',
    category: 'milestone',
    rarity: 'common',
    icon: Settings,
    color: '#CD7F32',
    colorEnd: '#B87333',
    requirement: '10 unlocks',
    criteria: { type: 'total_unlocks', value: 10 },
  },
  {
    id: 'the_operator',
    name: 'The Operator',
    description: 'Systems running at full capacity',
    visual: 'A sleek, chrome/silver mechanical heart',
    category: 'milestone',
    rarity: 'rare',
    icon: Heart,
    color: '#C0C0C0',
    colorEnd: '#A8A8A8',
    requirement: '50 unlocks',
    criteria: { type: 'total_unlocks', value: 50 },
  },
  {
    id: 'the_veteran',
    name: 'The Veteran',
    description: 'Battle-tested and proven',
    visual: 'A solid Gold pyramid or obelisk',
    category: 'milestone',
    rarity: 'epic',
    icon: Award,
    color: '#FFD700',
    colorEnd: '#FFA500',
    requirement: '100 unlocks',
    criteria: { type: 'total_unlocks', value: 100 },
  },
  {
    id: 'the_apex',
    name: 'The Apex',
    description: 'Peak performance achieved',
    visual: 'An Obsidian (black glass) diamond with purple inner light',
    category: 'milestone',
    rarity: 'legendary',
    icon: Gem,
    color: '#1F1F1F',
    colorEnd: '#8A2BE2',
    requirement: '500 unlocks',
    criteria: { type: 'total_unlocks', value: 500 },
  },
  {
    id: 'the_archivist',
    name: 'The Archivist',
    description: 'Knowledge meticulously preserved',
    visual: 'A glowing crystal shard or a futuristic data server rack',
    category: 'milestone',
    rarity: 'rare',
    icon: Database,
    color: '#06B6D4',
    colorEnd: '#0891B2',
    requirement: 'Saved 20 explanations',
    criteria: { type: 'saved_count', value: 20 },
  },
  {
    id: 'early_riser',
    name: 'Early Riser',
    description: 'Conquered the morning hours',
    visual: 'A digital sun rising over a retro-wave grid horizon',
    category: 'milestone',
    rarity: 'rare',
    icon: Sunrise,
    color: '#F59E0B',
    colorEnd: '#D97706',
    requirement: 'Attempt before 8:00 AM',
    criteria: { type: 'early_morning' },
  },
];

// Helper functions for badge operations
export function getBadgeById(id: string): Badge | undefined {
  return BADGE_DEFINITIONS.find(badge => badge.id === id);
}

export function getBadgesByCategory(category: BadgeCategory): Badge[] {
  return BADGE_DEFINITIONS.filter(badge => badge.category === category);
}

export function getBadgesByRarity(rarity: BadgeRarity): Badge[] {
  return BADGE_DEFINITIONS.filter(badge => badge.rarity === rarity);
}

export function getRarityColor(rarity: BadgeRarity): string {
  const colors = {
    common: '#9CA3AF',
    rare: '#60A5FA',
    epic: '#A855F7',
    legendary: '#FFD700',
  };
  return colors[rarity];
}

export function getRarityLabel(rarity: BadgeRarity): string {
  const labels = {
    common: 'Common',
    rare: 'Rare',
    epic: 'Epic',
    legendary: 'Legendary',
  };
  return labels[rarity];
}

File: iOS/ThinkFirst/Models/BadgeDefinitions.swift
import SwiftUI

extension BadgeCategory {
    var displayName: String {
        switch self {
        case .streak: return "Streaks"
        case .mastery: return "Mastery"
        case .milestone: return "Milestones"
        }
    }
}

extension BadgeRarity {
    var color: Color {
        switch self {
        case .common: return .gray
        case .rare: return .blue
        case .epic: return .purple
        case .legendary: return .yellow
        }
    }
}

struct BadgeDefinitions {
    static let allBadges: [Badge] = [
        // STREAK BADGES (6)
        Badge(
            id: "ignition",
            name: "Ignition",
            description: "The spark has been struck",
            visual: "A single matchstick striking a spark against a dark surface",
            category: .streak,
            rarity: .common,
            iconName: "flame.fill",
            color: "#FF6B35",
            colorEnd: "#FF4500",
            requirement: "3-day streak",
            criteria: BadgeCriteria(type: "streak", value: 3)
        ),
        Badge(
            id: "the_furnace",
            name: "The Furnace",
            description: "Heat is building, momentum is real",
            visual: "An industrial furnace door glowing orange with heat leaking out",
            category: .streak,
            rarity: .rare,
            iconName: "flame.fill",
            color: "#FF8C42",
            colorEnd: "#FF6B35",
            requirement: "7-day streak",
            criteria: BadgeCriteria(type: "streak", value: 7)
        ),
        Badge(
            id: "momentum",
            name: "Momentum",
            description: "The flywheel is spinning",
            visual: "A heavy flywheel or turbine spinning with green motion blur",
            category: .streak,
            rarity: .rare,
            iconName: "arrow.triangle.2.circlepath",
            color: "#00FF94",
            colorEnd: "#00CC75",
            requirement: "14-day streak",
            criteria: BadgeCriteria(type: "streak", value: 14)
        ),
        Badge(
            id: "blue_giant",
            name: "Blue Giant",
            description: "Burning with the hottest intensity",
            visual: "A massive star burning with intense blue/white light",
            category: .streak,
            rarity: .epic,
            iconName: "star.fill",
            color: "#00D9FF",
            colorEnd: "#0099CC",
            requirement: "30-day streak",
            criteria: BadgeCriteria(type: "streak", value: 30)
        ),
        Badge(
            id: "supernova",
            name: "Supernova",
            description: "Explosive learning energy",
            visual: "A star exploding in brilliant colors",
            category: .streak,
            rarity: .epic,
            iconName: "burst",
            color: "#FF00FF",
            colorEnd: "#8A2BE2",
            requirement: "60-day streak",
            criteria: BadgeCriteria(type: "streak", value: 60)
        ),
        Badge(
            id: "the_century",
            name: "The Century",
            description: "A hundred days of dedication",
            visual: "A golden monument with '100' carved in stone",
            category: .streak,
            rarity: .legendary,
            iconName: "crown.fill",
            color: "#FFD700",
            colorEnd: "#FFA500",
            requirement: "100-day streak",
            criteria: BadgeCriteria(type: "streak", value: 100)
        ),
        
        // MASTERY BADGES (7)
        Badge(
            id: "synapse",
            name: "Synapse",
            description: "First neural connection formed",
            visual: "Two neurons connecting with a spark of electricity",
            category: .mastery,
            rarity: .common,
            iconName: "brain.head.profile",
            color: "#9D4EDD",
            colorEnd: "#7B2CBF",
            requirement: "First mastery mode unlock",
            criteria: BadgeCriteria(type: "mastery_unlocks", value: 1)
        ),
        Badge(
            id: "neural_network",
            name: "Neural Network",
            description: "Building complex understanding",
            visual: "A web of interconnected neurons glowing softly",
            category: .mastery,
            rarity: .rare,
            iconName: "network",
            color: "#6A4C93",
            colorEnd: "#5A3A7C",
            requirement: "5 mastery mode unlocks",
            criteria: BadgeCriteria(type: "mastery_unlocks", value: 5)
        ),
        Badge(
            id: "deep_thinker",
            name: "Deep Thinker",
            description: "Diving into complex concepts",
            visual: "A person's silhouette with gears turning in their head",
            category: .mastery,
            rarity: .rare,
            iconName: "gearshape.2.fill",
            color: "#4361EE",
            colorEnd: "#3F37C9",
            requirement: "10 mastery mode unlocks",
            criteria: BadgeCriteria(type: "mastery_unlocks", value: 10)
        ),
        Badge(
            id: "perfectionist",
            name: "Perfectionist",
            description: "Flawless understanding achieved",
            visual: "A perfect crystal with light refracting through it",
            category: .mastery,
            rarity: .epic,
            iconName: "diamond.fill",
            color: "#00F5FF",
            colorEnd: "#0080FF",
            requirement: "Perfect score in mastery mode",
            criteria: BadgeCriteria(type: "perfect_score", value: nil)
        ),
        Badge(
            id: "night_owl",
            name: "Night Owl",
            description: "Learning burns bright in darkness",
            visual: "An owl perched on books under moonlight",
            category: .mastery,
            rarity: .rare,
            iconName: "moon.stars.fill",
            color: "#4C956C",
            colorEnd: "#2F5233",
            requirement: "Learn after 10 PM",
            criteria: BadgeCriteria(type: "late_night", value: nil)
        ),
        Badge(
            id: "early_bird",
            name: "Early Bird",
            description: "Dawn brings fresh insights",
            visual: "A bird singing at sunrise with books nearby",
            category: .mastery,
            rarity: .rare,
            iconName: "sunrise.fill",
            color: "#F77F00",
            colorEnd: "#D62828",
            requirement: "Learn before 6 AM",
            criteria: BadgeCriteria(type: "early_morning", value: nil)
        ),
        Badge(
            id: "the_polymath",
            name: "The Polymath",
            description: "Master of many domains",
            visual: "A tree with branches representing different subjects",
            category: .mastery,
            rarity: .legendary,
            iconName: "tree.fill",
            color: "#2D6A4F",
            colorEnd: "#1B4332",
            requirement: "25 mastery mode unlocks",
            criteria: BadgeCriteria(type: "mastery_unlocks", value: 25)
        ),
        
        // MILESTONE BADGES (7)
        Badge(
            id: "the_initiate",
            name: "The Initiate",
            description: "First step on the learning journey",
            visual: "A single footprint on a path leading into the distance",
            category: .milestone,
            rarity: .common,
            iconName: "figure.walk",
            color: "#06D6A0",
            colorEnd: "#048A81",
            requirement: "First unlock",
            criteria: BadgeCriteria(type: "total_unlocks", value: 1)
        ),
        Badge(
            id: "the_explorer",
            name: "The Explorer",
            description: "Venturing into new territories",
            visual: "A compass pointing toward unknown lands",
            category: .milestone,
            rarity: .common,
            iconName: "location.fill",
            color: "#118AB2",
            colorEnd: "#073B4C",
            requirement: "10 unlocks",
            criteria: BadgeCriteria(type: "total_unlocks", value: 10)
        ),
        Badge(
            id: "the_scholar",
            name: "The Scholar",
            description: "Dedicated to the pursuit of knowledge",
            visual: "Ancient scrolls and quills arranged on a wooden desk",
            category: .milestone,
            rarity: .rare,
            iconName: "book.fill",
            color: "#8B5A3C",
            colorEnd: "#6F4E37",
            requirement: "25 unlocks",
            criteria: BadgeCriteria(type: "total_unlocks", value: 25)
        ),
        Badge(
            id: "the_sage",
            name: "The Sage",
            description: "Wisdom flows through understanding",
            visual: "An ancient tree with glowing leaves of knowledge",
            category: .milestone,
            rarity: .rare,
            iconName: "leaf.fill",
            color: "#52B788",
            colorEnd: "#2D6A4F",
            requirement: "50 unlocks",
            criteria: BadgeCriteria(type: "total_unlocks", value: 50)
        ),
        Badge(
            id: "the_virtuoso",
            name: "The Virtuoso",
            description: "Mastery through persistent practice",
            visual: "Musical notes transforming into mathematical equations",
            category: .milestone,
            rarity: .epic,
            iconName: "music.note",
            color: "#E63946",
            colorEnd: "#A4161A",
            requirement: "100 unlocks",
            criteria: BadgeCriteria(type: "total_unlocks", value: 100)
        ),
        Badge(
            id: "the_legend",
            name: "The Legend",
            description: "Stories will be told of this dedication",
            visual: "A golden statue on a pedestal with rays of light",
            category: .milestone,
            rarity: .epic,
            iconName: "trophy.fill",
            color: "#FFB700",
            colorEnd: "#FF8500",
            requirement: "250 unlocks",
            criteria: BadgeCriteria(type: "total_unlocks", value: 250)
        ),
        Badge(
            id: "the_apex",
            name: "The Apex",
            description: "The pinnacle of learning achievement",
            visual: "A mountain peak touching the stars",
            category: .milestone,
            rarity: .legendary,
            iconName: "mountain.2.fill",
            color: "#7209B7",
            colorEnd: "#480CA8",
            requirement: "500 unlocks",
            criteria: BadgeCriteria(type: "total_unlocks", value: 500)
        )
    ]
}

// MARK: - Color Extension
extension Color {
    init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let a, r, g, b: UInt64
        switch hex.count {
        case 3: // RGB (12-bit)
            (a, r, g, b) = (255, (int >> 8) * 17, (int >> 4 & 0xF) * 17, (int & 0xF) * 17)
        case 6: // RGB (24-bit)
            (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
        case 8: // ARGB (32-bit)
            (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
        default:
            (a, r, g, b) = (1, 1, 1, 0)
        }

        self.init(
            .sRGB,
            red: Double(r) / 255,
            green: Double(g) / 255,
            blue:  Double(b) / 255,
            opacity: Double(a) / 255
        )
    }
}